cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java

```diff
@@ -0,0 +1,361 @@
+      = eventQueueProcessor.process(0, new LogFileOffsetTransformIterator(incomingEventQueue));
+
+    // none of the events were processed.
+    if (metadata.getTotalEventsProcessed() <= 0) {
+      return false;
+    }
+
+    unSyncedEvents += metadata.getTotalEventsProcessed();
+    // events were processed, so update the checkpoints
+    Checkpoint<LogBufferFileOffset> checkpoint = metadata.getCheckpoints().get(0);
+    MutableLogBufferCheckpoint mutableCheckpoint = checkpoints.get(0);
+    MutableLogBufferFileOffset offset = mutableCheckpoint.getOffset();
+    offset.setFileId(checkpoint.getOffset().getFileId());
+    offset.setFilePos(checkpoint.getOffset().getFilePos());
+    mutableCheckpoint.setMaxEventTs(checkpoint.getMaxEventTime());
+
+    return true;
+  }
+
+  /**
+   * This method is called from {@link ConcurrentLogBufferWriter} to push log events.
+   *
+   * @param events log events to be processed
+   */
+  public void processLogEvents(Iterator<LogBufferEvent> events) {
+    // Don't accept any log events if the pipeline is not running
+    if (!isRunning()) {
+      return;
+    }
+
+    while (events.hasNext()) {
+      try {
+        // This call will block ConcurrentLogBufferWriter thread until the queue has free space.
+        incomingEventQueue.put(events.next());
+      } catch (InterruptedException e) {
+        // Just ignore the exception and reset the flag
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Persists the checkpoints.
+   */
+  private void persistCheckpoints() {
+    try {
+      checkpointManager.saveCheckpoints(checkpoints);
+      LOG.debug("Checkpoint persisted for {} with {}", name, checkpoints);
+    } catch (Exception e) {
+      // Just log as it is non-fatal if failed to save checkpoints
+      OUTAGE_LOG.warn("Failed to persist checkpoint for pipeline {}.", name, e);
+    }
+  }
+
+  /**
+   * Sync the appender and persists checkpoints if it is time.
+   *
+   * @return delay in millisecond till the next sync time.
+   */
+  private long trySyncAndPersistCheckpoints(long currentTimeMillis) {
+    if (unSyncedEvents == 0) {
+      return config.getCheckpointIntervalMillis();
+    }
+    if (currentTimeMillis - config.getCheckpointIntervalMillis() < lastCheckpointTime) {
+      return config.getCheckpointIntervalMillis() - currentTimeMillis + lastCheckpointTime;
+    }
+
+    // Sync the appender and persists checkpoints
+    try {
+      context.sync();
+      // Only persist if sync succeeded. Since persistCheckpoints never throw, it's ok to be inside the try.
+      persistCheckpoints();
+      lastCheckpointTime = currentTimeMillis;
+      metricsContext.gauge("last.checkpoint.time", lastCheckpointTime);
+      unSyncedEvents = 0;
+      LOG.debug("Events synced and checkpoint persisted for {}", name);
+    } catch (Exception e) {
+      OUTAGE_LOG.warn("Failed to sync in pipeline {}. Will be retried.", name, e);
```

Comment:
What does the retrying?

Reply:
It will be done by run method https://github.com/cdapio/cdap/pull/11080/files#diff-0def6fa6acd833eaa7edbbb45acf188dR103

Reply:
In that case, it's better if this method throws the exception and the logging/handling is done by the caller that is actually doing the retrying.